packages/accounts-password/password_server.js

```diff
@@ -758,7 +833,97 @@ Meteor.methods({verifyEmail: function (token) {
   );
 }});
 
+/**
+ * @summary Add an email address for a user. Use this instead of directly
+ * updating the database. The operation will fail if there is a different user
+ * with an email only differing in case. If the specified user has an existing
+ * email only differing in case however, we replace it.
+ * @locus Server
+ * @param {String} userId The ID of the user to update.
+ * @param {String} newEmail A new email address for the user.
+ * @param {Boolean} [verified] Optional - whether the new email address should
+ * be marked as verified. Defaults to false.
+ */
+Accounts.addEmail = function (userId, newEmail, verified) {
+  check(userId, NonEmptyString);
+  check(newEmail, NonEmptyString);
+  check(verified, Match.Optional(Boolean));
+
+  if (_.isUndefined(verified)) {
+    verified = false;
+  }
+
+  var user = Meteor.users.findOne(userId);
+  if (!user)
+    throw new Meteor.Error(403, "User not found");
+
+  // Allow users to change their own email to a version with a different case
+  var caseInsensitiveRegExp =
+    new RegExp('^' + Meteor._escapeRegExp(newEmail) + '$', 'i');
+
+  var didUpdateOwnEmail = _.any(user.emails, function(email, index) {
+    if (caseInsensitiveRegExp.test(email.address)) {
+      Meteor.users.update({
+        _id: user._id,
+        'emails.address': email.address
+      }, {$set: {
+        'emails.$.address': newEmail,
+        'emails.$.verified': verified
+      }});
+      return true;
+    }
+
+    return false;
+  });
+
+  if (didUpdateOwnEmail) {
+    return;
```

Comment:
I was a little confused at first as to why we don't need something like the "call `checkForCaseInsensitiveDuplicates` both before and after the Mongo operation" pattern here. But then I understood that other than modifying Mongo directly (in which case you're on your own), there's no way for another user to add `newEmail` while this user has it in another case.

Maybe a comment explaining this? Up to you, it may be clearer than I make it seem.